Initial description of applications and identity providers (#1)
tomleavy/mimi-arch-draft: draft-aegis-mimi-arch-latest.md

' +-------+     +----------------------+     +----------------+ '
' +-------+     +----------------------+       |                '
' | Dave  | --- | Application Server 4 |-------+                '
' +-------+     +----------------------+                        '
'                                                               '
+ - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - +
~~~

# Gateway Overview

The MIMI Gateway is designed to be a hybrid of AS and DS functionality as
defined by `MLS Architecture`. It has an internal interface for
applications within its domain to federate with each other, as well as an
external interface that allows for cross-domain federation. A gateway acts as a
permissioned database of the following information:

* Unique identifiers of clients that can be queried by application-defined tags.
* Queryable queues of MLS Key Packages indexed by identifier, protocol version,
cipher suite and a custom label.
* MLS proposals, commits, welcome messages and ratchet trees indexed by MLS
`group_id` and `epoch_id`.

## Identity Providers

An identity provider helps ensure a consistent AS behavior between federating

> LS applications. Internally an identity provider MUST implement the following

Reviewer: ```suggestion
MLS applications. Internally an identity provider MUST implement the following
```

set of functionality. 

* Declare support for a specific set of MLS credential types.
* Validate an identity based on its public key, credential, the current set of
  group context extensions, and opaque application context. If validation is
  done in the context of a commit, group context extensions MUST include any
  changes due to a GroupContextExtensions proposal.
* Uniquely identify a client based on its identity to ensure that the same
  identity is not added to a MLS group multiple times.
* Determine if one identity is a valid successor of another in order to verify
  if an Update proposal, or a remove proposal within an external commit should
  be allowed. Unless otherwise specified, a client is a valid successor of

>   another client if their client identifiers are equal.

Reviewer 2: For updates (and commits) we need valid_successor() to make sure its the same client.

But for Remove via External Join we need something else: same_account(Identity oldID, Identity newID) which checks that the creds belong to the same account (but allows them to be for different clients). So we also need account_identifierIdentity Iid) -> VEc<u8>.

Another time we need same_acount() is when I'm managing my own devices in a group (although I'm not a moderator for the group). Say, I loose my cellphone and then want to remove it from a group  using my tablet. Normally this remove isn't allowed coz i'm not a mod, but since remover and removee are on the same account its actually  OK.

An example interface of an identity provider is as follows:

~~~
struct {
    HPKEPublicKey public_key;
    Credential credential;
} Identity;

struct {
    opaque context<V>
} ApplicationContext;

interface {
    fn supported_credentials() -> [CredentialType];

>     fn validate(Identity identity, Extension group_context_ext[],

Reviewer 2: By forcing the context that comes from the MLS session to be the group_context_ext arent we forcing all MLS extensions that want to contribute to validation contexts to insert their data into the group_context? If so is that intended?

* Leaf and entity identifiers are equal to the bare assertion of identity
  provided by a BasicCredential.

## X.509 Identity Provider

The X.509 identify provider allows for a hierarchical identity that is based
upon a certificate chain. x509 is the only credential type supported.
Certificate chains MUST be validated according to the rules in RFC 5280 using
the trust roots agreed upon by the (TODO: Some sort of extension dealing with
trust root negotiation).

~~~
struct {
   uint64 msg_timestamp;
} X509ApplicationContext;

struct {
    X509IdentityRange range;
} X509Parameters;
~~~

### X.509 Identifiers

Calculating a client's identifier is done by iterating through a subset of X.509
subject fields found within the client's certificate chain. The range of
certificates in the chain to reference as part of the calculation is defined as
starting with `start` values from the leaf and ending with
`end` values from the leaf. A single unique identifier is calculated by
hashing the Common Name value found in each certificate within the reference
range using the current group's ciphersuite hash function.

~~~
struct {
   uint8 start;
   uint8 end<0..255>;
} X509IdentityRange;

> ~~~

Reviewer 2: ```suggestion
```

fn client_identifier(HashFunction hasher, CertificateChain cert_chain,
X509IdentityRange range)
{
    for certificate in cert_chain[range.start..=range.end] {
        let common_name = certificate.subject.common_name;

        if (!common_name) {
            throw "Common name is required";
        }

        hasher.update(common_name);
    }

    return hasher.finalize();
}

> ~~~

Reviewer: The disadvantage of hashing is 1) looking at ID you can't tell much, e.g., what's the domain, 2) identities are different for different ciphersuites. What if the identifier is simply a list, serialized in some way? This could be easily turned into e.g. Rohan's URI scheme.

Reviewer 2: Suggestion to make this compatible with Rohan's URI scheme:
 - Subject Names of certs in identity range can't include following reserved characters: @ / # $ :
 - Client Identifier := Subject Names of certs in identity range concatenated with : as separator between them.

This way we get:
 - Client IDs are unambiguous. E.g. client_identifier(Cert1.subjName = "AB" + Cert2.SubjName = "C") != client_identifier(Cert1.SubjName = "A" + Cert2.Subjname = "BC").
 - compatible with Rohan's URI. You can just drop in our client_identifier() output as the client/device component in a URI conforming to Rohan's scheme.

> 

Reviewer 2: I suggest we start the hash with a constant string input like "clientID". Meanwhile for account_identifier() we start the hash with a different constant string inut like "accountID". That domain seperats IDs for us and avoids any chance of colliding IDs between different types of entities.

_identity_ may change (for instance, if they change phone number).
Further, each client is (optionally) assigned a set of tags. Each tag
represents a grouping of clients, for example:
1. A tag `account_id:alice@org` represents all clients belonging to an
   account `alice@org`.
2. Tags `moderator` or `clearance:top_secret` represent all clients with
   given access rights.

Tags of a given client are provided by the application. The service
allows the application to search for clients by tags. For example,
searching by tag 1. above allows to list all clients owned by
`alice@org`, which is a typical scenario in messaging applications where
Bob invites a multi-device account of Alice. Searching by a tag 2.
above allows to add an arbitrary moderator, or all clients owned by
entities with the right clearance.

## Identity Registration

When a client registers with an application, the application registers
the client with the identity service using the following request:

~~~
HTTP POST /apps/{appId}/clients 

Input:

{
    clientIdentity: Base64(IdentityProvider.client_identity),
    tags: [String]
}

Output:

{
    clientId: UUIDv4
}

~~~

> In the above figure, the returned `client_id` is the new unique

Reviewer: ```suggestion
In the above figure, the returned `clientId` is the new unique
```

# Introduction

In order to achieve cross-application federation between a set of MLS applications there must be a common set of APIs available that implement the basic AS/DS requirements as defined by `MLS Architecture`. To facilitate the creation of groups asynchronously, MLS clients may have a need to establish identity, credentials and a queue of key packages. Once groups are established, it is useful for clients to have access to a common repository for welcome messages, pending proposals, commit messages, and ratchet trees. 

> In this document we describe a MIMI Gateway API that can be used to provide intra domain and cross domain federation between MLS applications.

Reviewer: I'm missing how identity provider fits into the bigger picture? Currently it appears out of the blue in the next section. So I would

1. make this more general, like "we describe the API of AS and DS components that can be used to provide intra domain and cross domain federation between MLS applications. These components include  MIMI Gateway API and Identity Provider".
2. in the operational context below, place the provider in the bigger picture, something like "To ensure consistent behavior, all actors that execute MLS, including all clients and Gateways, must agree on its AS components (e.g., credential types used by MLS). The AS-related functionality that actors need to agree on is collected in the Identity Provider"

Author: Ran out of time today, can you try to address that?

# Operational Context 

> A basic federation scenario consists of a bi-directional data flow between actors. Clients

Reviewer: ```suggestion
A basic federation scenario consists of a bi-directional data flow between three types of actors: clients, application servers and MIMI gateways. Clients

```

communicate with application servers, application servers
communicate with gateway services, and gateway services
communicate with other gateway services.

~~~
+ - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - +
'                       Domain A                                '
'                                                               '
' +-------+     +----------------------+     +----------------+ '
' | Alice | --- | Application Server 1 | --- | MIMI Gateway A |-'-----+

> ' +-------+     +----------------------+     +----------------+ '     |

Reviewer: Would it help to show that a gateway is useful even if an app doesn't federate with anything?

Author: I think separate diagram, but yeah that makes sense

Reviewer: Created #1

' |  Bob  | --- | Application Server 2 |-------+                '     |
' +-------+     +----------------------+                        '     |
'                                                               '     |
+ - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - +     |
                                                                      |
+ - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - +     |
'                       Domain B                                '     |
'                                                               '     |
' +-------+     +----------------------+     +----------------+ '     |
' | Carol | --- | Application Server 3 | --- | MIMI Gateway B |-'-----+ 
' +-------+     +----------------------+     +----------------+ '
' +-------+     +----------------------+       |                '
' | Dave  | --- | Application Server 4 |-------+                '
' +-------+     +----------------------+                        '
'                                                               '
+ - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - +
~~~

# Gateway Overview

> The MIMI Gateway is designed to be a hybrid of AS and DS functionality has

Reviewer: ```suggestion
The MIMI Gateway is designed to be a hybrid of AS and DS functionality as

```

defined by `MLS Architecture`. It has an internal interface for
applications within its domain to federate with each other, as well as an
external interface that allows for cross-domain federation. A gateway acts as a
permissioned database of the following information:

> * Unique identities that can be grouped by tags.

Reviewer: ```suggestion
* Unique identifiers of clients that can be queried by application-defined tags.
```

> * Queryable queues of MLS Key Packages indexed by identity, protocol version, cipher suite, custom labels.

Reviewer: ```suggestion
* Queryable queues of MLS Key Packages indexed by identifier, protocol version, cipher suite and a custom label.
```

> * MLS Group proposals, commits, welcome messages and ratchet trees indexed by epoch.

Reviewer: ```suggestion
* MLS proposals, commits, welcome messages and ratchet trees indexed by MLS `group_id` and `epoch_id`.
```

## Identity Providers

An identity provider helps ensure a consistent AS behavior between federating
MLS applications. Internally an identity provider MUST implement the following
set of functionality. 

* Declare support for a specific set of MLS credential types.

> * Validate a credential given a public key that is stored along side that

Reviewer: ```suggestion
* Validate a credential given a public key that is stored alongside that

```

  credential in an MLS Leaf and a timestamp that can be used to determine
  validity of credentials that may expire (Ex. X509 Certificate credentials).
* Uniquely identify a Leaf based on its identity to ensure that the same
  identity is not added to a MLS group multiple times.
* Uniquely identity the entity that controls a specific leaf.
* Determine if an identity is controlled by the same entity as another identity.

An example interface of an identity provider is as follows:

~~~
struct {
    HPKEPublicKey public_key;
    Credential credential;
} Identity;

fn supported_credentials() -> [CredentialType];
fn validate(Identity identity, u64 time) -> bool;
fn leaf_identifier(Identity identity) -> Vec<u8>;

> fn entity_identifier(Identity identity) -> Vec<u8>;

Reviewer: I really hate bringing this up... The RFC ended up saying
> the LeafNode in the path field of the external commit MUST meet the same criteria as would the LeafNode in an Update for the removed leaf

So it looks like we only need one thing here. Maybe the leaf cred doesn't matter at all; it just holds an ephemeral signature key?

Author: Lets chat about this, I'm not sure what you are getting at. The primary reason for having that identifier is to make sure the same client doesn't get added twice to a convo.

fn valid_successor(Identity predecessor, Identity successor) -> bool {
    return self.entity_identifier(predecessor) == self.entity_identifier(successor);

> }

Reviewer: Should we remove this function given we have entity identifiers?

Author: I think we can probably do that, but we still need to specify how to do this for external commit right?

Reviewer: I misunderstood. Thanks!

~~~

Each identity provider has a unique value that identifies its behavior along
with opaque parameters that help synchronize various options across
applications. 

~~~
uint16 IdentityProviderType;

struct {
    IdentityProviderType type;
    opaque parameters<V>;
} IdentityConfiguration;
~~~

> ## Basic Identity Provider 

Reviewer: Any reason to use this outside of tests?

Author: I figured the MLS spec has it, so why not have it as a reference?

above allows to add an arbitrary moderator, or all clients owned by
entities with the right clearance.

## Identity Registration
When a client registers with an application, the application registers
the client with the identity service using the following query:

// TODO: This should be described as an HTTP route.

~~~
u16 IdentityQueryType = 1;

struct {
    opaque entity_identifier<V>;
    opaque tag<V>;

> } RegisterIdentityQuery;

Reviewer: Should there also be RegisterIdentityResponse?

Author: Yes, also, this should be HTTP + JSON. We need to figure out how we want to generally display request response. I asked around, there is no standard way of doing it.
